src/net/raw/devices/devices-macos.c

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2015 click2stream, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <net/if.h>
+#include <net/if_arp.h>
+#include <net/if_dl.h>
+#include <netinet/in.h>
+#include <ifaddrs.h>
+
+#include "devices.h"
+#include "devices-utils.h"
+#include "utils.h"
+
+static int get_mac_address(const char* dname, unsigned char* buffer) {
+    struct ifaddrs* iflist;
+    struct ifaddrs* cur;
+
+    if (getifaddrs(&iflist) == 0) {
+        for (cur = iflist; cur; cur = cur->ifa_next) {
```

Comment:
Why do you call `getifaddrs()` here and look up the device again? You already have the device when you get to the `get_device_info()` function (it's passed as argument from `net_find_devices()`. You just need to check here that `sa_family == AF_LINK` (or you can check it in `net_find_devices()`).

Reply:
You're right. It can be simplified and `getifaddrs` doesn't need to be called twice.
The problem is that `getifaddrs()` returns records like this.
```
"eth0", AF_LINK, "11:22:33:44:55:66"
"eth0", AF_INET, "192.168.1.1"
```
So now I iterate through the array N*N times.
First I find IPv4 addresses and for the second time, when I try to obtain the MAC address for the device name.

Reply:
ok, I didn't realize that you cannot get MAC address and IP address at once